src/character.cpp

```diff
@@ -652,19 +652,19 @@ void Character::i_rem_keep_contents( const int pos )
     }
 }
 
-bool Character::i_add_or_drop(item& it, int qty) {
+bool Character::i_add_or_drop( item& it, int qty ) {
     bool retval = true;
     bool drop = false;
-    inv.assign_empty_invlet(it);
-    for (int i = 0; i < qty; ++i) {
-        if ( !drop && ( !can_pickWeight( it, !get_option<bool>( "DANGEROUS_PICKUPS" ) )
+    inv.assign_empty_invlet( it );
+    for( int i = 0; i < qty; ++i ) {
+        if( !drop && ( !can_pickWeight( it, !get_option<bool>( "DANGEROUS_PICKUPS" ) )
                       || !can_pickVolume( it ) ) ) {
             drop = true;
         }
-        if( drop ) {
+        if( drop || it.made_of( LIQUID ) ) {
```

Comment:
This can be optimized a bit (by moving relatively "heavy" operations out of the loop): 
```c++
     bool retval = true;
     bool drop = it.made_of( LIQUID );
     bool add = it.is_gun() || !it.has_flag("IRREMOVABLE");
     ....
     for( int i = 0; i < qty; ++i ) {
         drop |= !can_pickWeight( ...

         if( drop ) {
             ...
         } else if( add ) {
             ...
         }
    }
```